File: tasks/02-fix-the-bug/python/buggy_counter.py
# buggy_counter.py

import threading
import time

_current = 0
# using the thread Lock
_t_lock = threading.Lock()


def next_id():
    """Returns a unique ID, incrementing the global counter."""
    # may be get the same time and duplicate value when run in thread
    global _current
    # using the thread Lock that lock the thread access and then release Lock
    # using with to guarantee release the lock
    with _t_lock:
        value = _current
        time.sleep(0)
        _current += 1
        return value

#Solution : Using the thread Lock the controle the thread lock and release
# Taken time : 20 minutes

File: tasks/02-fix-the-bug/go/buggy_counter.go
package counter

import (
	"sync"
	"time"
)

var mu sync.Mutex

var current int64

func NextID() int64 {
	// lock the gorounting
	mu.Lock()
	// after unlock function exit
	defer mu.Unlock()

	id := current
	time.Sleep(0)
	current++
	return id
}
